httpserver/server.go

```diff
@@ -36,18 +38,39 @@ const (
 	certName          = "tls.crt"
 	keyName           = "tls.key"
 	readHeaderTimeout = 5 * time.Second
+
+	// cacheTTL is the default time-to-live for the cache entry.
+	cacheTTL     = 5 * time.Minute
+	cacheMaxSize = 100
 )
 
-type (
-	Server struct {
-		Address       string
-		Router        *mux.Router
-		GetExecutor   config.GetExecutor
-		Context       context.Context
-		CertDirectory string
-		CaCertFile    string
+type Server struct {
+	Address       string
+	Router        *mux.Router
+	GetExecutor   config.GetExecutor
+	Context       context.Context
+	CertDirectory string
+	CaCertFile    string
+
+	keyMutex keyMutex
+	// cache is a thread-safe expiring lru cache which caches external data item indexed
+	// by the subject
+	cache *simpleCache
+}
+
+// keyMutex is a thread-safe map of mutexes, indexed by key.
+type keyMutex struct {
+	locks sync.Map
+}
+
+// Lock locks the mutex for the given key, and returns a function to unlock it.
+func (m *keyMutex) Lock(key string) func() {
+	v, _ := m.locks.LoadOrStore(key, &sync.Mutex{})
+	v.(*sync.Mutex).Lock()
+	return func() {
+		v.(*sync.Mutex).Unlock()
 	}
-)
+}
```

Comment:
Instead of this, we can also use https://pkg.go.dev/golang.org/x/sync/singleflight that'll ensure only a single call is inflight for the same subject

Reply:
Which one do you think best suits our needs?

Reply:
I'll switch it to `singleflight`. We can do a custom lock implementation if at some point what `singleflight` offers isn't sufficient for our scenario.

Reply:
Both are the same. `singleflight` is still `v0.1.0`, so we can keep the key mutex for now. If there are newer feature in `singleflight` we want in the future, it can be switched!